proxy/src/fully_qualified_authority.rs

```diff
@@ -0,0 +1,236 @@
+use bytes::BytesMut;
+
+use std::ascii::AsciiExt;
+use std::net::IpAddr;
+use std::str::FromStr;
+
+use http::uri::Authority;
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct FullyQualifiedAuthority(Authority);
+
+impl FullyQualifiedAuthority {
+    /// Normalizes the name according to Kubernetes service naming conventions.
+    /// Case folding is not done; that is done internally inside `Authority`.
+    ///
+    /// This assumes the authority is syntactically valid.
+    pub fn new(authority: &Authority, default_namespace: Option<&str>,
+               default_zone: Option<&str>)
+               -> FullyQualifiedAuthority {
+        // Don't change IP-address-based authorities.
+        if IpAddr::from_str(authority.host()).is_ok() {
+            return FullyQualifiedAuthority(authority.clone())
+        };
+
+        // TODO: `Authority` doesn't have a way to get the serialized form of the
+        // port, so do it ourselves.
+        let (name, colon_port) = {
+            let authority = authority.as_str();
+            match authority.rfind(':') {
+                Some(p) => authority.split_at(p),
+                None => (authority, ""),
+            }
+        };
+
+        // A fully qualified name ending with a dot is normalized by removing the
+        // dot and doing nothing else.
+        if name.ends_with('.') {
+            let authority = authority.clone().into_bytes();
+            let normalized = authority.slice(0, authority.len() - 1);
+            return FullyQualifiedAuthority(Authority::from_shared(normalized).unwrap());
```

Comment:
We should probably strive to have `unwrap`s be descriptive `expect`s instead.

Reply:
(> We should probably strive to have unwraps be descriptive expects instead.

I switched it to `expect()`, though I question whether it's really better.